File: 114/wordladder.hs
import System.Environment
import Control.Applicative
import Data.String.Utils
import Data.Ord
import Data.List

dictFile = "selected_four-letter_words.txt"

-- Command line interface

usage = "\nInvalid syntax.\n\nUsage:\n\n"

    ++ "./wordladder list <word>\n"
    ++ "    - List all the words that can be made from <word> by changing one letter.\n\n"

    ++ "./wordladder most\n"
    ++ "    - Show the top 10 biggest word ladders.\n\n"

    ++ "./wordladder chain <n> <word>\n"
    ++ "    - Show the number of words that can be reached, starting from <word>, in <n> or fewer steps.\n\n"

    ++ "Examples:\n\n"

    ++ "./wordladder list best\n"
    ++ "./wordladder most\n"
    ++ "./wordladder chain 3 best\n"

main = do
    args <- getArgs
    dict <- lines <$> replace "\r" "" <$> readFile dictFile
    case args of
        "list":word:[] -> putStrLn $ fmtList $ wordLadder word dict
        "most":_ -> putStrLn $ fmtList $ flattenPairs $ mostLadderable dict
        "chain":steps:start:[] -> putStrLn $ show $ length $ wordChain (read steps) (start:[]) dict
        _ -> putStrLn usage

fmtList list = foldl1 (\x acc-> x ++ "\n" ++ acc) list

flattenPairs list = map (\p -> show (fst p) ++ " - " ++ show (snd p)) list

-- Word Ladder

wordLadder :: String -> [String] -> [String]
wordLadder word dict = filter (oneLetterOff word) dict

oneLetterOff :: String -> String -> Bool
oneLetterOff word1 word2 = 1 == (length $ filter not $ (zipWith (==) word1 word2))

-- Most Ladderable

mostLadderable dict =
    let ladderLengthPairs = map (\w -> (w, length (wordLadder w dict))) dict
    in
    take 10 $ reverse $ sortBy (comparing snd) ladderLengthPairs

-- Word Chain

wordChain :: Int -> [String] -> [String] -> [String]
wordChain 0 current _ = current
wordChain steps current dict =
    let nextLevelWords = nub $ concatMap (\x -> wordLadder x dict) current
    in
    nub (current ++ wordChain (steps-1) nextLevelWords dict)

File: 114/scala/WordLadder.scala
import scala.io.Source._

object WordLadder {

    val dictFile = "../selected_four-letter_words.txt"

    // Word Ladder

    def wordLadder(word: String, dict: Seq[String]): Seq[String] = 
        dict.filter(oneLetterOff(word, _))


    private def oneLetterOff(word1: String, word2: String): Boolean = 
        (word1, word2).zipped.map(_ == _).filter(x => !x).length == 1

    // Most Ladderable

    def mostLadderable(num: Int, dict: Seq[String]): Seq[(String, Int)] = {
        val wordNumberPairs = for (w <- dict) yield (w, wordLadder(w, dict).length)
        wordNumberPairs.sortBy(_._2).reverse.take(num)
    }

    // Word Chain

    def wordChain(steps: Int, current: String, dict: Seq[String]): Set[String] = {
        var finalSet = Set(current)
        var next = nextLevelOfWords(finalSet, dict)

        for (i <- 0 until steps) {
            finalSet = finalSet ++ next
            next = nextLevelOfWords(next, dict)
        }

        return finalSet
    }

    private def nextLevelOfWords(current: Set[String], dict: Seq[String]): Set[String] = 
        current.flatMap(w => wordLadder(w, dict))


    // Command line interface

    val usage = "usage info"

    def fmtList(list: Seq[String]): String = list.mkString("\n")

    def fmtPairs(pairs: Seq[(String, Int)]): String = pairs.flatMap(p => p._1 + " - " + p._2 + "\n").mkString

    def main(args: Array[String]): Unit = {
        val dict = fromFile(dictFile).getLines.toSeq

        args toList match {
            case "list" :: word :: Nil            => println(fmtList(wordLadder(word, dict)))
            case "top" :: num :: Nil              => println(fmtPairs(mostLadderable(num.toInt, dict)))
            case "chain" :: steps :: start :: Nil => println(wordChain(steps.toInt, start, dict).size)
            case _                                => println(usage)
        }
    }
}

